Preserve user specified attributes and methods

bzinodev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38830 Preserve user specified attributes and methods**

This patch enables to preserve user specified attributes or non forward
methods. The API:
  _freeze_module(Module, ["a", "version"])

Differential Revision: [D21957316](https://our.internmc.facebook.com/intern/diff/D21957316)

bzinodev committed:
Preserve user specified attributes and methods

This patch enables to preserve user specified attributes or non forward
methods. The API:
  _freeze_module(Module, ["a", "version"])

[ghstack-poisoned]

bzinodev committed:
Update on "Preserve user specified attributes and methods"


This patch enables to preserve user specified attributes or non forward
methods. The API:
  _freeze_module(Module, ["a", "version"])

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -14,7 +14,33 @@ namespace {
+      return false;
+    };
+
+    // forward is preserved by default.
```
I don't know if we want this - if this is true there will be no way to remove the forward function. I think the signature should just be `freeze_module(module, List[str])` and then `freeze_module(module)` which will freeze all but forward.

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -36,17 +62,24 @@ class AttributePropagator {
-    cleanupFrozenModule(graph);
+    for (auto function : preservedMethods_) {
+      auto graph = function->graph();
+      optimizeSubGraphs(graph, applyInline);
```
This is sequentially runs `propagateAttributes` on each method, which records mutable attributes and then inlines the non-mutable ones. What we want instead is to first collect the set of all mutable attributes on all methods, and then inline them.

eellison commented on test/jit/test_freezing.py:
```diff
@@ -891,3 +891,47 @@ def test_freeze_module_detach_gradient(self):
+        self.assertTrue(fm.hasattr("a"))
+        self.assertFalse(fm.hasattr("b"))
+
+    def test_freeze_module_with_user_preserved_method(self):
```
Can you add a test like:

        class Module(nn.Module):
            def __init__(self):
                super(Module, self).__init__()
                self.a = torch.tensor([1.1])
                self.b = torch.tensor([2.2])

            def forward(self, x):
                self.b[0] += 10
                return self.a + self.b

            @torch.jit.export
            def modify_a(self, x):
                self.a[0] += 10 
                return self. b + self.a

        m = torch.jit.script(Module())
        m.eval()
        fm = torch._C._freeze_module(m._c, ["modify_a"])
        FileCheck().check('prim::GetAttr[name="a"]').run(fm.forward.graph)
        FileCheck().check('prim::GetAttr[name="b"]').run(fm.modify_a.graph)

eellison reviewed (commented):
Looks good - one question about the API and one question about the order we evaluate recordMutableAttrs in

bzinodev committed:
Update on "Preserve user specified attributes and methods"


This patch enables to preserve user specified attributes or non forward
methods. The API:
  _freeze_module(Module, ["a", "version"])

[ghstack-poisoned]

bzinodev commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -219,8 +219,11 @@ void initJITBindings(PyObject* module) {
           "_freeze_module",
-          [](Module& module) { return freeze_module(module); },
-          py::arg("module"))
+          [](Module& module, std::vector<std::string> preservedAttrs) {
```
I think no. forward is expected to be always preserved

bzinodev committed:
Update on "Preserve user specified attributes and methods"


This patch enables to preserve user specified attributes or non forward
methods. The API:
  _freeze_module(Module, ["a", "version"])

[ghstack-poisoned]

bzinodev committed:
Update on "Preserve user specified attributes and methods"


This patch enables to preserve user specified attributes or non forward
methods. The API:
  _freeze_module(Module, ["a", "version"])

[ghstack-poisoned]

bzinodev committed:
Update on "Preserve user specified attributes and methods"


This patch enables to preserve user specified attributes or non forward
methods. The API:
  _freeze_module(Module, ["a", "version"])

Differential Revision: [D21957316](https://our.internmc.facebook.com/intern/diff/D21957316)

[ghstack-poisoned]

eellison reviewed (approved):
Code looks good to me, please update the one test I commented to be a little more confident it is working as expected

bzinodev committed:
Update on "Preserve user specified attributes and methods"


This patch enables to preserve user specified attributes or non forward
methods. The API:
  _freeze_module(Module, ["a", "version"])

Differential Revision: [D21957316](https://our.internmc.facebook.com/intern/diff/D21957316)

[ghstack-poisoned]
